### PR TITLE
Bugfix include cache generations on querier response

### PR DESCRIPTION
**Pull request by owen-d**
We check the cache number (incremented by deletes) after executing a query with the value returned by the queriers which executed it. If they’re different, we know we can’t safely cache. However, we’re never actually returning a number from the queriers. This means that as soon as this cache number is incremented by a delete, it’ll disable the results cache for that tenant completely.

I suspect this has been a latent bug since we forked cortex (goes back to missing some code from this [PR](https://github.com/cortexproject/cortex/pull/2279) into cortex from @sandeepsukhani  back in the day!). It just showed itself once we started using deletes.

### CODE REVIEW

File: pkg/loki/modules.go [hunk 1]
```diff
@@ -364,6 +364,7 @@ func (t *Loki) initQuerier() (services.Service, error) {
 
 	httpMiddleware := middleware.Merge(
 		httpreq.ExtractQueryMetricsMiddleware(),
+		serverutil.CacheGenNumberHeaderSetterMiddleware(t.cacheGenerationLoader),
```
Comment (sandeepsukhani) [inline pkg/loki/modules.go:367]: we can add a check for `supportIndexDeleteRequest` to see if we need to add this middleware or not?

**Review by sandeepsukhani — APPROVED** [review]
Nice catch! LGTM!

**Review by chaudum — COMMENTED** [review]
Great find! Can we verify in a test that the query response contains the `ResultsCacheGenNumberHeaderName` header value if index deletes are supported?

**Comment by owen-d** [comment]
> Great find! Can we verify in a test that the query response contains the ResultsCacheGenNumberHeaderName header value if index deletes are supported?

Yes, I think we can (perhaps in a larger integration test), but I'll leave it as a followup opportunity.

**Pull request merged**